src/Sylius/Bundle/ApiBundle/Validator/Constraints/AddingEligibleProductVariantToCartValidator.php

```diff
@@ -85,7 +86,31 @@ public function validate($value, Constraint $constraint)
             return;
         }
 
-        if (!$this->availabilityChecker->isStockSufficient($productVariant, $value->quantity)) {
+        /** @var OrderInterface|null $cart */
+        $cart = $this->orderRepository->findCartByTokenValue($value->getOrderTokenValue());
+        Assert::notNull($cart);
+
+        $units = $cart->getItemUnits();
+
+        $sameOrderItemsCounter = 0;
+
+        if ($productVariant->isTracked() && !$units->isEmpty()) {
+            foreach ($units as $unit) {
+                /** @var OrderItemInterface $orderItem */
+                $orderItem = $unit->getOrderItem();
+
+                /** @var ProductVariantInterface */
+                $variant = $orderItem->getVariant();
+
+                $code = $variant->getCode();
+
+                if ($value->productVariantCode = $code) {
+                    ++$sameOrderItemsCounter;
+                }
+            }
+        }
+
+        if (!$this->availabilityChecker->isStockSufficient($productVariant, $value->quantity + $sameOrderItemsCounter)) {
```

Comment:
You can create a private method for providing a quantity of given product in the cart, sth like:
```
    private function getExistingCartItemQuantityFromCart(OrderInterface $cart, ProductVariantInterface $productVariant): int
    {
        foreach ($cart->getItems() as $existingCartItem) {
            if ($existingCartItem-> getVariant()->equals($productVariant)) {
                return $existingCartItem->getQuantity();
            }
        }

        return 0;
    }
```
and then
```suggestion
        if (!$this->availabilityChecker->isStockSufficient($productVariant, $value->quantity + $this->getExistingCartItemQuantityFromCart($cart, $productVariant))) {
```